jolt-sdk/macros/src/lib.rs

```diff
@@ -1,298 +1,363 @@
-                let arg_name = &pat_ident.ident;
-                let arg_type = &ty;
-
-                let arg_fetch = quote! {
-                    let x = input_slice[0];
-                    let (#arg_name, input_slice) = jolt_sdk::postcard::take_from_bytes::<#arg_type>(input_slice).unwrap();
-                };
+struct MacroBuilder {
+    attr: AttributeArgs,
+    func: ItemFn,
+    func_args: Vec<(Ident, Box<Type>)>,
+}
 
-                args.push(arg_fetch);
-                arg_types.push(arg_type.clone());
-            } else {
-                panic!("cannot parse arg");
-            }
-        } else {
-            panic!("cannot parse arg");
+impl MacroBuilder {
+    fn new(attr: AttributeArgs, func: ItemFn) -> Self {
+        let func_args = Self::get_func_args(&func);
+        Self {
+            attr,
+            func,
+            func_args,
         }
     }
 
-    // TODO: ensure that input slice hasn't overflown
-    let check_input_len = quote! {};
+    fn build(&self) -> TokenStream {
+        let build_fn = self.make_build_fn();
+        let execute_fn = self.make_execute_function();
+        let preprocess_fn = self.make_preproccess_func();
+        let prove_fn = self.make_prove_func();
+        let main_fn = self.make_main_func();
+
+        quote! {
+            #build_fn
+            #execute_fn
+            #preprocess_fn
+            #prove_fn
+            #main_fn
+        }
+        .into()
+    }
 
-    let block = &input_fn.block;
-    let block = quote! {let to_return = (|| -> _ { #block })();};
+    fn make_build_fn(&self) -> TokenStream2 {
+        let fn_name = self.get_func_name();
+        let build_fn_name = Ident::new(&format!("build_{}", fn_name), fn_name.span());
+        let prove_output_ty = self.get_prove_output_type();
+
+        let input_names = self.func_args.iter().map(|(name, _)| name);
+        let input_types = self.func_args.iter().map(|(_, ty)| ty);
+        let inputs = &self.func.sig.inputs;
+        let preprocess_fn_name = Ident::new(&format!("preprocess_{}", fn_name), fn_name.span());
+        let prove_fn_name = syn::Ident::new(&format!("prove_{}", fn_name), fn_name.span());
+        let imports = self.make_imports();
+        
+
+        quote! {
+            #[cfg(not(feature = "guest"))]
+            pub fn #build_fn_name() -> (
+                impl Fn(#(#input_types),*) -> #prove_output_ty,
+                impl Fn(jolt_sdk::Proof) -> bool
+            ) {
+                #imports
+                let (program, preprocessing) = #preprocess_fn_name();
+                let program = std::rc::Rc::new(program);
+                let preprocessing = std::rc::Rc::new(preprocessing);
+
+                let program_cp = program.clone();
+                let preprocessing_cp = preprocessing.clone();
+
+                let prove_closure = move |#inputs| {
+                    let program = (*program).clone();
+                    let preprocessing = (*preprocessing).clone();
```

Comment:
is there any way to avoid these clones? otherwise these will be included in our reported benchmark times

Reply:
We do need them as the prove and verify functions take ownership of their inputs.

Fortunately if you look a couple lines above I moved the cloned data into a `Rc` which makes the cloning costs insignificant.

Reply:
ah gotcha